ignite/contrib/handlers/time_profilers.py

```diff
@@ -437,3 +438,395 @@ def odict_to_str(d: Mapping):
+
+
+class HandlersTimeProfiler:
+    """
+    HandlersTimeProfiler can be used to profile the handlers,
+    data loading and data processing times.
+
+    Examples:
+
+    .. code-block:: python
+
+        from ignite.contrib.handlers import HandlersTimeProfiler
+
+        trainer = Engine(train_updater)
+
+        # Create an object of the profiler and attach an engine to it
+        profiler = HandlersTimeProfiler()
+        profiler.attach(trainer)
+
+        @trainer.on(Events.EPOCH_COMPLETED)
+        def log_intermediate_results():
+            profiler.print_results(profiler.get_results())
+
+        trainer.run(dataloader, max_epochs=3)
+
+        profiler.write_results('path_to_dir/time_profiling.csv')
+
+    """
+
+    def __init__(self):
+        self._dataflow_timer = Timer()
+        self._processing_timer = Timer()
+        self._event_handlers_timer = Timer()
+
+        self.dataflow_times = None
+        self.processing_times = None
+        self.event_handlers_times = None
+
+        # Events to profile
+        self._events = [
+            Events.STARTED,
+            Events.EPOCH_STARTED,
+            Events.EPOCH_COMPLETED,
+            Events.ITERATION_STARTED,
+            Events.ITERATION_COMPLETED,
+            Events.GET_BATCH_STARTED,
+            Events.GET_BATCH_COMPLETED,
+            Events.COMPLETED,
+        ]
+
+        # Events which are triggered only once
+        self._global_events = [Events.STARTED, Events.COMPLETED]
+        # Events which are triggered on each EPOCH
+        self._epoch_events = [Events.EPOCH_STARTED, Events.EPOCH_COMPLETED]
+        # Events which are triggered on each ITERATION
+        self._iteration_events = [
+            Events.ITERATION_STARTED,
+            Events.ITERATION_COMPLETED,
+            Events.GET_BATCH_STARTED,
+            Events.GET_BATCH_COMPLETED,
+        ]
+
+    @staticmethod
+    def _get_callable_name(handler: Callable):
+        # get name of the callable handler
+        return handler.__qualname__ if hasattr(handler, "__qualname__") else handler.__class__.__name__
+
+    def _timeit_global_events(self, handler: Callable, event: Events):
+        # wrapper for profiling handlers attached to global events
+        self._event_handlers_timer.reset()
+        func, args, kwargs = handler
+        func(*args, **kwargs)
+        t = self._event_handlers_timer.value()
+        hname = self._get_callable_name(func)
+        self.event_handlers_times[event][hname][0] = t
+
+    def _timeit_epoch_events(self, handler: Callable, engine: Engine, event: Events, *args, **kwargs):
+        # wrapper for profiling handlers attached to epoch events
+        self._event_handlers_timer.reset()
+        func, args, kwargs = handler
+        func(*args, **kwargs)
+        e = engine.state.epoch - 1
+        t = self._event_handlers_timer.value()
+        hname = hname = self._get_callable_name(func)
+        self.event_handlers_times[event][hname][e] = t
```

Comment:
Maybe, instead of adding to a slot we can append to a list. There is another point, a bit unrelated to this handler but which may impact the behaviour. For example, if user adds a handler attached to a filtered event like `Events.EPOCH_COMPLETED(every=3)`. Currently, this `every=3` filtering is translated to the handler: decorated handler is executed, falls into condition that we should or not execute user's original handler. So, for example, my handler takes in average 1s to execute, the times we'll have then: 
- epoch 1, T ~ 0s (time to filter)
- epoch 2, T ~ 0s (time to filter)
- epoch 3, T ~ 1s (time to filter + original handler)
- epoch 4, T ~ 0s (time to filter)
...

This is something to check and think about. Maybe, we have to add a threshold and do not append times smaller then this threshold.

Reply:
Okay. Let me see if I can fine tune the threshold for "every" and "once" event filters.

Reply:
For custom filters I am not sure how to handle them.